### DISCUSSION TITLE
Drop SCOPE attribute?

### DISCUSSION THREAD

**Discussion by staabm** [rectorphp/rector#7924] [General]
similar to https://github.com/rectorphp/rector/issues/7849 I wonder whether we could get rid of the `AttributeKey::SCOPE`-attribute we attach on a lot of nodes, by consequently using `ScopeAwarePhpRectorInterface` and `refactorWithScope` where `Scope` is required?

of course we would need to pass the `Scope` object thru some layers and helpers, but I think that it is done in phpstan and also in a lot of already existing functionality

As I understood it, rector is constantly changing the scope while working thru the codebase and since a lot of nodes reference all these scope-objects involved, my thesis would be that php-src is not able to free the scope-objects memory, as they are still referenced in a lot of places.

is there are reason why sometimes the scope is retrieved from a node attribute instead of `refactorWithScope`?
or should `refactorWithScope` work everywhere and it is just not done yet?

any opinions?

**Comment by TomasVotruba** [comment]
Hi, this would be quite useful improvement :+1: In the end, it would be best to have simply refactor() method with both node and scope, like PHPStan. One way over optional is better design.

Now it would be useful to have real data about how much scope attributes can we remove. Having one by one small PRs like we do parent/next/previous node.
